contracts/ExitHandler.sol

```diff
@@ -29,42 +29,392 @@ contract ExitHandler is DepositHandler {
+    });
+    for (uint8 i =0; i < transferTx.ins.length; i++){
+      TxLib.Input memory ins = transferTx.ins[i];
+      TxLib.Output memory out = transferTx.outs[ins.outpoint.pos];
+      inputs[i] = LimboIn({
+        amount: out.value,
+        owner: out.owner, 
+        isPegged: false, 
+        exitable: true,
+        challenge: {}
+      });
+    }
+    
+    bytes32 inTxHash = keccak256(inTxData);
+    bytes32 utxoId = bytes32(uint256(_outputIndex) << 120 | uint120(uint256(inTxHash)));
+    uint256 priority;
+    uint32 height;
+    uint32 timestamp;
+
+    // priority based on youngestInput
+    if (_youngestInputProof.length > 0) {
+      (height, timestamp) = bridge.periods(_youngestInputProof[0]);
+      require(timestamp > 0, "The referenced period was not submitted to bridge");
+    }
+    bytes32 inputTxHash;
+    uint64 txPos;
+    (txPos, inputTxHash,) = TxLib.validateProof(96, _youngestInputProof);
+
+    if (isNft(outs.color)) {
+      priority = (nftExitCounter << 128) | uint128(uint256(utxoId));
+      nftExitCounter++;
+    } else {      
+      priority = getERC20ExitPriority(timestamp, utxoId, txPos);
+    }
+    limboExits[utxoId] = LimboExit({
+      input: inputs,
+      output: outputs,
+      finalized: false,
+      exitor: msg.sender,
+      isValid: true
+    });
+
+    emit LimboExitStarted(
+      inTxHash, 
+      outs.color
+    );
+    tokens[outs.color].insert(priority);
+
+    return utxoId;
+  }
+
+  }
+
+  function joinLimboExit(
+    bytes32 exitId, uint8 _index) 
+    public payable {
+    require(msg.value >= piggybackStake, "Not enough ether sent to join the exit");
+
+    address owner = msg.sender;
+    LimboExit memory limboExit = limboExits[exitId];
+
+    if (limboExit.input[_index].owner == owner){
+      // input is piggybacking
+      require(limboExit.input[_index].isPegged = false, "Already joined the exit");
+
+      limboExit.input[_index].isPegged = true;
+    } else if (limboExit.output[_index].owner == owner) {
+      // output is piggybacking
+      require(limboExit.output[_index].isPegged = false, "Already joined the exit");
+
+      limboExit.output[_index].isPegged = true;
+    }
+  
+  }
+
+  function challengeLimboExitByInclusionProof(
+    bytes32 exitId, bytes memory inTxData,
+    uint8 InputNo, bytes32[] memory _youngestInputProof, bytes32[] memory  _inclusionProof) 
+    public payable {
+    require(msg.value >= challengeStake, "Not enough ether sent to challenge exit");
+    LimboExit memory limboExit = limboExits[exitId];
+    
+    bytes32 inTxHash = keccak256(inTxData);
+    require(limboExit.isValid == true);
+    TxLib.Tx memory transferTx = TxLib.parseTx(inTxData);
+    
+    bytes32 txHash;
+    bytes memory txData;
+    uint64 txPos;
+    (txPos, txHash, txData) = TxLib.validateProof(32 * (_youngestInputProof.length + 2) + 64, _inclusionProof);
+
+    require(inTxHash == txHash, "Invalid inclusion proof");
+
+    limboExit.isValid = false;
```

Comment:
This function is missing the actual check against limbo exit. Having a unit test will help.